chemprop/hyperopt_utils.py

```diff
@@ -0,0 +1,210 @@
+from chemprop.args import HyperoptArgs
+import os
+import pickle
+from typing import List, Dict
+import csv
+import json
+
+from hyperopt import Trials
+
+from chemprop.constants import HYPEROPT_SEED_FILE_NAME
+from chemprop.utils import makedirs
+
+def merge_trials(trials: Trials, new_trials_data: List[Dict]) -> Trials:
+    """
+    Merge a hyperopt trials object with the contents of another hyperopt trials object.
+
+    :param trials: A hyperopt trials object containing trials data, organized into hierarchical dictionaries.
+    :param trials_data: The contents of a hyperopt trials object, `Trials.trials`.
+    :return: A hyperopt trials object, merged from the two inputs.
+    """
+    max_tid = 0
+    if len(trials.trials) > 0:
+        max_tid = max([trial['tid'] for trial in trials.trials])
+
+    for trial in new_trials_data:
+        tid = trial['tid'] + max_tid + 1 #trial id needs to be unique among this list of ids.
+        hyperopt_trial = Trials().new_trial_docs(
+                tids=[None],
+                specs=[None],
+                results=[None],
+                miscs=[None])
+        hyperopt_trial[0] = trial
+        hyperopt_trial[0]['tid'] = tid
+        hyperopt_trial[0]['misc']['tid'] = tid
+        for key in hyperopt_trial[0]['misc']['idxs'].keys():
+            hyperopt_trial[0]['misc']['idxs'][key] = [tid]
+        trials.insert_trial_docs(hyperopt_trial) 
+        trials.refresh()
+    return trials
+
+
+def load_trials(dir_path: str, previous_trials: Trials = None) -> Trials:
+    """
+    Load in trials from each pickle file in the hyperopt checkpoint directory.
+    Checkpoints are newly loaded in at each iteration to allow for parallel entries
+    into the checkpoint folder by independent hyperoptimization instances.
+
+    :param dir_path: Path to the directory containing hyperopt checkpoint files.
+    :param previous_trials: Any previously generated trials objects that the loaded trials will be merged with.
+    :return: A trials object containing the merged trials from all checkpoint files.
+    """
+
+    # List out all the pickle files in the hyperopt checkpoint directory
+    hyperopt_checkpoint_files = [os.path.join(dir_path,path) for path in os.listdir(dir_path) if '.pkl' in path]
+
+    # Load hyperopt trials object from each file
+    loaded_trials = Trials()
+    if previous_trials is not None:
+        loaded_trials = merge_trials(loaded_trials,previous_trials.trials)
+
+    for path in hyperopt_checkpoint_files:
+        with open(path,'rb') as f:
+            trial = pickle.load(f)
+            loaded_trials = merge_trials(loaded_trials, trial.trials)
```

Comment:
Why is the empty `Trials()` in `loaded_trials` merged in? I.e. I am not sure why line 58 and 59 is necessary. Wouldn't it be enough to have something like:
```
if previous_trials is not None:
    loaded_trials=previous_trials.trials
else:
    loaded_trials=Trials()
```

(maybe I am missing something, though)

Reply:
So what I found is that when you merge data into a Trials object to make it a larger Trials object, that is an inplace operation. So if you didn't create a new blank Trials instance here to merge into, then anything you added to `previous_trials` would be added to the Trials object that you fed in as an argument to the function. This keeps it from modifying that object outside the function.

Since I want to load in checkpoints anew every time anew every time, I don't want to accumulate that data anywhere. I just want it in a new fresh object every time. And since the same Trials object is being fed in as previous_trials every iteration, it wouldn't just accumulate the loaded trials but would accumulate many duplicate copies of them, which would then be counted against the total number of iterations.

Reply:
Ah, that is super interesting and makes sense, thank you!